### ISSUE TITLE
unexpected object replacement instead of update using kubernetes_manifest resource

### ISSUE THREAD

**Issue by riuvshyn** [hashicorp/terraform-provider-kubernetes#1841]
I am using `kubernetes_manifest` to create and manage rancher `Cluster` k8s object, when I am trying to change `Cluster` configuration by modifying anything under `spec.rkeConfig.machineGlobalConfig` terraform provider wants to replace Cluster object which is not acceptable behaviour.

Why kubernetes terraform provider decided that anychange to `spec.rkeConfig.machineGlobalConfig` forces replacement? 

### Terraform Version, Provider Version and Kubernetes Version

```
Terraform version:
Kubernetes provider version: 2.13.1
Kubernetes version: 1.24.4
```

### Affected Resource(s)
`kubernetes_manifest`

### Terraform Configuration Files
<details>
  <summary>tf resource</summary>
  
  ```hcl

resource "kubernetes_manifest" "rancher_cluster" {
  wait {
    condition {
      type   = "Created"
      status = "True"
    }
    condition {
      type   = "RKECluster"
      status = "True"
    }
  }
  manifest = {
    "apiVersion" = "provisioning.cattle.io/v1"
    "kind"       = "Cluster"
    "metadata" = {
      "labels" = {
        "provider.cattle.io"             = "rke2"
      }
      "name"      = local.cluster_name
      "namespace" = "fleet-default"
    }
    "spec" = {
      "defaultClusterRoleForProjectMembers" = "user"
      "enableNetworkPolicy"      = false
      "kubernetesVersion"        = var.rke2_version
      "localClusterAuthEndpoint" = {}
      "rkeConfig" = {
        "etcd" = {
          "s3" = {
            "bucket"   = "my-bucket"
            "endpoint" = "s3.eu-central-1.amazonaws.com"
            "folder"   = "etcd"
            "region"   = "eu-central-1"
          }
          "snapshotRetention"    = 12
          "snapshotScheduleCron" = "0 */6 * * *"
        }
        "machineGlobalConfig" = {
          "cloud-provider-name" = "aws"
          "cluster-cidr"   = "100.64.0.0/13"
          "cluster-dns"    = "100.64.0.10"
          "cluster-domain" = "cluster.local"
          "cni"            = "none"
          "disable" = [
            "rke2-ingress-nginx",
            "rke2-metrics-server",
            "rke2-canal",
          ]
          "disable-cloud-controller" = false
          "kube-apiserver-arg" = [
            "allow-privileged=true",
            "anonymous-auth=false",
            "feature-gates=${join(",", ["CustomCPUCFSQuotaPeriod=true"])}",
            "api-audiences=https://${module.oidc.oidc_issuer_domain},https://kubernetes.default.svc.cluster.local,rke2",
            "audit-log-maxage=90",
            "audit-log-maxbackup=10",
            "audit-log-maxsize=500",
            "audit-log-path=/var/log/k8s-audit/audit.log",
            "audit-policy-file=/etc/kubernetes-wise/audit-policy.yaml",
            "authorization-mode=Node,RBAC",
            "bind-address=0.0.0.0",
            "enable-admission-plugins=PodSecurityPolicy,NodeRestriction",
            "event-ttl=1h",
            "kubelet-preferred-address-types=InternalIP,Hostname,ExternalIP",
            "profiling=false",
            "request-timeout=60s",
            "runtime-config=api/all=true",
            "service-account-key-file=/etc/kubernetes-wise/service-account.pub",
            "service-account-lookup=true",
            "service-account-issuer=https://${module.oidc.oidc_issuer_domain}",
            "service-account-signing-key-file=/etc/kubernetes-wise/service-account.key",
            "service-node-port-range=30000-32767",
            "shutdown-delay-duration=60s",
            "tls-cipher-suites=${local.tls_cipher_suites}",
            "tls-min-version=${local.tls_min_version}",
            "v=2",
          ]
          "kube-apiserver-extra-mount" = [
            "/etc/kubernetes-wise:/etc/kubernetes-wise:ro",
            "/var/log/k8s-audit:/var/log/k8s-audit:rw",
          ]
          "kube-controller-manager-arg" = [
            "allocate-node-cidrs=true",
            "attach-detach-reconcile-sync-period=1m0s",
            "bind-address=0.0.0.0",
            "configure-cloud-routes=false",
            "feature-gates=${join(",", ["CustomCPUCFSQuotaPeriod=true"])}",
            "leader-elect=true",
            "node-monitor-grace-period=2m",
            "pod-eviction-timeout=220s",
            "profiling=false",
            "service-account-private-key-file=/etc/kubernetes-wise/service-account.key",
            "use-service-account-credentials=true",
            "terminated-pod-gc-threshold=12500",
            "tls-cipher-suites=${local.tls_cipher_suites}",
            "tls-min-version=${local.tls_min_version}",
          ]
          "kube-controller-manager-extra-mount" = [
            "/etc/kubernetes-wise:/etc/kubernetes-wise:ro",
          ]
          "kube-proxy-arg" = [
            "conntrack-max-per-core=131072",
            "conntrack-tcp-timeout-close-wait=0s",
            "metrics-bind-address=0.0.0.0",
            "proxy-mode=iptables",
          ]
          "kube-scheduler-arg" = [
            "bind-address=0.0.0.0",
            "secure-port=10259",
            "profiling=false",
            "leader-elect=true",
            "v=2",
            "tls-cipher-suites=${local.tls_cipher_suites}",
            "tls-min-version=${local.tls_min_version}",
          ]
          "kubelet-arg" = [
            "feature-gates=${join(",", ["CustomCPUCFSQuotaPeriod=true"])}",
            "config=/etc/kubernetes-wise/kubelet.yaml",
            "exit-on-lock-contention=true",
            "lock-file=/var/run/lock/kubelet.lock",
            "pod-infra-container-image=docker-k8s-gcr-io.tw.ee/pause:3.1",
            "register-node=true",
            "tls-cipher-suites=${local.tls_cipher_suites}",
            "tls-min-version=${local.tls_min_version}",
            "v=4",
          ]
          "profile"                 = "cis-1.6"
          "protect-kernel-defaults" = false
          "service-cidr" = "100.64.0.0/13"
        }
        "registries" = {
          "configs" = {
            "docker.tw.ee" = {
              "insecureSkipVerify" = true
            }
          }
          "mirrors" = local.container_registry_mirrors
        }
        "upgradeStrategy" = {
          "controlPlaneConcurrency" = "1"
          "controlPlaneDrainOptions" = {
            "deleteEmptyDirData"              = true
            "disableEviction"                 = false
            "enabled"                         = var.drain_nodes_before_upgrade
            "force"                           = false
            "gracePeriod"                     = 0
            "ignoreDaemonSets"                = true
            "skipWaitForDeleteTimeoutSeconds" = 600
            "timeout"                         = 10800
          }
          "workerConcurrency" = var.upgrade_worker_concurrency
          "workerDrainOptions" = {
            "deleteEmptyDirData"              = true
            "disableEviction"                 = false
            "enabled"                         = var.drain_nodes_before_upgrade
            "force"                           = false
            "gracePeriod"                     = 0
            "ignoreDaemonSets"                = true
            "skipWaitForDeleteTimeoutSeconds" = 600
            "timeout"                         = 10800
          }
        }
      }
    }
  }
}
```
</details>

### Steps to Reproduce
any change to `spec.rkeConfig.machineGlobalConfig` for some reason forces replace, replace is not acceptable on any `spec` attribute change.

### Expected Behavior
k8s object update happens in place

### Actual Behavior
```
  # module.cluster.kubernetes_manifest.rancher_cluster must be replaced
+/- resource "kubernetes_manifest" "rancher_cluster" {
      ~ manifest = {
          ~ spec       = {
              ~ rkeConfig                           = {
                  ~ machineGlobalConfig   = {
                      ~ protect-kernel-defaults             = true -> false
                        # (16 unchanged elements hidden)
                    } # forces replacement
                    # (5 unchanged elements hidden)
                }
                # (4 unchanged elements hidden)
            }
            # (3 unchanged elements hidden)
        }
```

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by jbg** [comment]
Probably the CRD has no schema, so in-place updates aren't possible. If this is the case then you get a warning after apply, like this:

```
│ Warning: This custom resource does not have an associated OpenAPI schema.
│ 
│   with kubernetes_manifest.example,
│   on example.tf line 1, in resource "kubernetes_manifest" "example":
│  1: resource "kubernetes_manifest" "example" {
│ 
│ We could not find an OpenAPI schema for this custom resource. Updates to this resource will cause a forced replacement.
```

**Comment by riuvshyn** [comment]
@jbg hello, thanks for looking in to this, no I don't get that warning message.
but in the CRD:
```
                  machineGlobalConfig:
                    nullable: true
                    type: object
                    x-kubernetes-preserve-unknown-fields: true
```

**Comment by alexsomesan** [comment]
@riuvshyn Please share the version of your rancher install and the cluster you are running on.

**Comment by alexsomesan** [comment]
TL;DR; Any CRD that contains `x-kubernetes-preserve-unknown-fields` will behave like this. 

The long-winded explanation is the following:
CRD authors using that flag basically renounce specifying schema. Without schema information, the provider is forced to replace the resource in order to satisfy Terraform's requirements of resource structure invariance (a resource cannot modify it's declared structure during its lifecycle).

Please raise an issue with Rancher and encourage them to fully specify the schema of their CRDs.

**Comment by riuvshyn** [comment]
rancher version is 2.6.8
